File: 260_SingleNumberIII.java
// Given an array of numbers nums, in which exactly two elements appear only once and all the other elements appear exactly twice. Find the two elements that appear only once.

// For example:

// Given nums = [1, 2, 1, 3, 2, 5], return [3, 5].

//Idea: Bit Manipulation : xor.
//      a^a = 0, a^0 = a, a^b = b^a
//      Given nums = [a,b,c,d,a,b]. 
//      int xor = a^b^c^d^a^b = a^a^b^b^c^d = c^d
//      int flag = xor & ~(xor - 1).
//      xor & ~(xor - 1) returns the first 1 in xor.
//      And 1 in xor means that bit is different in c and d.
//      Say c = 1001, d = 1111. xor = c^d = 0110. 
//      flag = xor & ~(xor - 1) = 0110 & ~(0101) = (0110) & (1010) = 0010
//      we can divide the numbers into two groups using the flag, one contains c and the other contains d.

public class Solution {
    public int[] singleNumber(int[] nums) {
        int[] res = new int[2];
        int xor = 0;
        res[0] = 0;
        res[1] = 0;
        for(int i = 0; i < nums.length; i++){
            xor ^= nums[i];
        }
        int flag = xor & ~(xor - 1);
        for(int num : nums){
            if((num & flag) == 0){
                res[0] ^= num;
            }else{
                res[1] ^= num;
            }
        }
       return res;
    }
}

File: PYTHON/200-299/260_Single_Number_III.py
#
#Given an array of numbers nums, in which exactly two elements appear only once and all the other elements appear exactly twice. Find the two elements that appear only once.
#
#
#For example:
#
#
#Given nums = [1, 2, 1, 3, 2, 5], return [3, 5].
#
#
#Note:
#
#The order of the result is not important. So in the above example, [5, 3] is also correct.
#Your algorithm should run in linear runtime complexity. Could you implement it using only constant space complexity?
#
#
#
#Credits:Special thanks to @jianchao.li.fighter for adding this problem and creating all test cases.
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        ret = [0,0]
        xor = reduce(operator.xor,nums)
        xor &= ~(xor - 1)
        for num in nums:
            if num & xor == 0 : ret[0] ^= num
            else : ret[1] ^= num
        return ret
        